src/sknnr/datasets/_base.py

```diff
@@ -0,0 +1,156 @@
+from __future__ import annotations
+
+import csv
+import sys
+from dataclasses import dataclass
+from importlib import resources
+from typing import TYPE_CHECKING, TextIO, Union
+
+import numpy as np
+from numpy.typing import NDArray
+
+if TYPE_CHECKING:  # pragma: no cover
+    import pandas as pd
+
+DATA_MODULE = "sknnr.datasets.data"
+
+
+@dataclass
+class Dataset:
+    ids: Union[NDArray[np.int64], "pd.DataFrame"]
+    data: Union[NDArray[np.float64], "pd.DataFrame"]
+    target: Union[NDArray[np.float64], "pd.DataFrame"]
+    frame: Union[None, "pd.DataFrame"]
```

Comment:
Great question! These are [forward references](https://peps.python.org/pep-0484/#forward-references), which as you guessed were needed because `pandas` won't be imported at runtime. However, because I put in `from __future__ import annotations` in order to be able to use built-in types (`tuple` instead of `typing.Tuple`), that seems to remove the need for this.

I'll do a little more research to make sure there aren't any caveats for older Python versions, but I think I can safely remove the quotes.

Reply:
Interesting.  Does it relate to [this note](https://docs.python.org/3/library/typing.html#typing.TYPE_CHECKING)?  I'll admit to still being a bit confused, but as long as you have it wired, I'm happy.

Reply:
Exactly, good find! That note and the related [PEP 563](https://peps.python.org/pep-0563/) and [PEP 649](https://peps.python.org/pep-0649) seem to confirm that as long as we keep the `__future__` import, Python won't try to evaluate those types during runtime if we remove the quotes.

If you're looking for some thrilling reading material, the [History of Annotations](https://peps.python.org/pep-0649/#a-history-of-annotations) in PEP 649 goes into a lot of depth on why this feature exists.

Reply:
Awesome, I'm going to dig in!